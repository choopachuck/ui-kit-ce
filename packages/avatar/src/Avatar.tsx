'use client'

/* eslint-disable react/prop-types */ // TODO: косячит eslint при типизации пропсов компонента через дженерики React.forwardRef

import * as React from 'react'
import { createUseStyles, clsx } from '@v-uik/theme'
import { useClassList } from '@v-uik/hooks'
import { useImageLoaded } from './hooks'
import { AvatarProps } from './types'
import { FallbackIcon } from './assets'
import { AvatarKind, AvatarSize } from './constants'
import { AvatarGroupContext } from './AvatarGroupContext'

const useStyles = createUseStyles(
  (theme) => {
    const avatarCircleStyle = {
      borderTopLeftRadius: theme.comp.avatar.shapeBorderRadiusTopLeftCircle,
      borderTopRightRadius: theme.comp.avatar.shapeBorderRadiusTopRightCircle,
      borderBottomLeftRadius:
        theme.comp.avatar.shapeBorderRadiusBottomLeftCircle,
      borderBottomRightRadius:
        theme.comp.avatar.shapeBorderRadiusBottomRightCircle,
    }
    const avatarRoundedStyle = {
      borderTopLeftRadius: theme.comp.avatar.shapeBorderRadiusTopLeftRounded,
      borderTopRightRadius: theme.comp.avatar.shapeBorderRadiusTopRightRounded,
      borderBottomLeftRadius:
        theme.comp.avatar.shapeBorderRadiusBottomLeftRounded,
      borderBottomRightRadius:
        theme.comp.avatar.shapeBorderRadiusBottomRightRounded,
    }

    const avatarSquareStyle = {
      borderTopLeftRadius: theme.comp.avatar.shapeBorderRadiusTopLeftSquare,
      borderTopRightRadius: theme.comp.avatar.shapeBorderRadiusTopRightSquare,
      borderBottomLeftRadius:
        theme.comp.avatar.shapeBorderRadiusBottomLeftSquare,
      borderBottomRightRadius:
        theme.comp.avatar.shapeBorderRadiusBottomRightSquare,
    }

    const avatarDefaultSizeStyle = {
      fontSize: 16,
      lineHeight: '24px',
      letterSpacing: 0.15,
      '& $icon': {
        '& svg': {
          height: 24,
          width: 24,
        },
      },
    }

    return {
      root: {
        fontFamily: theme.comp.avatar.typographyFontFamily,
        fontWeight: theme.comp.avatar.typographyFontWeight,
        display: 'inline-flex',
        alignItems: 'center',
        justifyContent: 'center',
        flexDirection: 'column',
        color: theme.comp.avatar.colorText,
      },
      colorDefault: {
        backgroundColor: theme.comp.avatar.colorBackground,
      },
      image: {
        height: '100%',
        width: '100%',
        objectFit: 'cover',
      },
      circle: {
        ...avatarCircleStyle,
        '& $image': {
          ...avatarCircleStyle,
        },
        '&$withBorder::after': {
          ...avatarCircleStyle,
        },
        '& $fallback': {
          ...avatarCircleStyle,
        },
      },
      rounded: {
        ...avatarRoundedStyle,
        '& $image': {
          ...avatarRoundedStyle,
        },
        '&$withBorder::after': {
          ...avatarRoundedStyle,
        },
        '& $fallback': {
          ...avatarRoundedStyle,
        },
      },
      square: {
        ...avatarSquareStyle,
        '& $image': {
          ...avatarSquareStyle,
        },
        '&$withBorder::after': {
          ...avatarSquareStyle,
        },
        '& $fallback': {
          ...avatarSquareStyle,
        },
      },
      customSize: {
        ...avatarDefaultSizeStyle,
      },
      extraSmall: {
        fontSize: 12,
        lineHeight: '16px',
        letterSpacing: 0.4,
        height: 24,
        width: 24,
        '& $icon': {
          '& svg': {
            height: 16,
            width: 16,
          },
        },
      },
      small: {
        fontSize: 14,
        lineHeight: '16px',
        letterSpacing: 0.25,
        height: 32,
        width: 32,
        '& $icon': {
          '& svg': {
            height: 20,
            width: 20,
          },
        },
      },
      medium: {
        height: 40,
        width: 40,
        ...avatarDefaultSizeStyle,
      },
      large: {
        fontSize: 20,
        lineHeight: '24px',
        letterSpacing: 0.15,
        height: 48,
        width: 48,
        '& $icon': {
          '& svg': {
            height: 32,
            width: 32,
          },
        },
      },
      extraLarge: {
        fontSize: 24,
        lineHeight: '32px',
        letterSpacing: 0.35,
        height: 64,
        width: 64,
        '& $icon': {
          '& svg': {
            height: 32,
            width: 32,
          },
        },
      },
      icon: {
        display: 'flex',
        height: '100%',
        width: '100%',
        alignItems: 'center',
        justifyContent: 'center',
      },
      withShadow: {
        boxShadow: theme.comp.avatar.elevationShadow,
      },
      withBorder: {
        position: 'relative',
        '&::after': {
          content: '""',
          position: 'absolute',
          height: '100%',
          width: '100%',
          border: `1px solid ${theme.comp.avatar.colorBorder}`,
        },
      },
      fallback: {
        display: 'flex',
        height: '100%',
        width: '100%',
        alignItems: 'center',
        justifyContent: 'center',
        overflow: 'hidden',
      },
    }
  },
  { name: 'avatar' }
)

const getAttribute = <T extends unknown = string>(
  attr: keyof JSX.IntrinsicElements['img'],
  props?: JSX.IntrinsicElements['img'],
  defaultValue?: T
) => {
  return (props?.[attr] || defaultValue) as T
}

/**
 * Функция генерации текста для аватара, который состоит из каждой первой буквы первых двух слов из свойства `alt`
 */
const getCaptionFromAlt = ({ alt, imgProps }: AvatarProps) => {
  const altSplitted = getAttribute('alt', imgProps, alt)?.split(' ') || ['', '']

  return `${altSplitted?.[0]?.[0] ?? ''}${altSplitted?.[1]?.[0] ?? ''}`
}

export const Avatar = React.forwardRef<HTMLDivElement, AvatarProps>(
  (
    {
      classes,
      size: sizeProp,
      kind: kindProp,
      color: colorProp,
      children,
      src: srcProp,
      alt,
      imgProps,
      icon,
      withBorder: withBorderProp,
      withShadow: withShadowProp,
      style,
      fallback = <FallbackIcon />,
      className,
      ...rest
    },
    ref
  ) => {
    let content: React.ReactNode = null
    const classesList = useStyles()
    const classesMap = useClassList(classesList, classes)
    const src = getAttribute('src', imgProps, srcProp)
    const imageLoadedStatus = useImageLoaded({
      ...imgProps,
      src,
    })
    const avatarGroupContext = React.useContext(AvatarGroupContext)

    const size = sizeProp || avatarGroupContext?.size || AvatarSize.md
    const kind = kindProp || avatarGroupContext?.kind || AvatarKind.circle

    const color = colorProp || avatarGroupContext?.color
    const withBorder = withBorderProp || avatarGroupContext?.withBorder

    const withShadow = withShadowProp || avatarGroupContext?.withShadow

    const customAvatarSize = typeof size === 'number' ? size : undefined

    const avatarClassName = clsx(classesMap.root, className, {
      // size
      [classesMap.extraSmall]: size === AvatarSize.xs,
      [classesMap.small]: size === AvatarSize.sm,
      [classesMap.medium]: size === AvatarSize.md,
      [classesMap.large]: size === AvatarSize.lg,
      [classesMap.extraLarge]: size === AvatarSize.xl,
      [classesMap.customSize]: !!customAvatarSize,
      // kind
      [classesMap.circle]: kind === AvatarKind.circle,
      [classesMap.rounded]: kind === AvatarKind.rounded,
      [classesMap.square]: kind === AvatarKind.square,
      // default background
      [classesMap.colorDefault]: !color,
      // border
      [classesMap.withBorder]: withBorder,
      // shadow
      [classesMap.withShadow]: withShadow,
    })

    const imgBaseProps = {
      ...imgProps,
      className: clsx(classesMap.image, imgProps?.className),
    }

    const hasImage =
      imageLoadedStatus !== 'error' && (!!src || !!imgProps?.srcSet)

    const inlineStyle = {
      background: color,
      height: customAvatarSize,
      width: customAvatarSize,
      ...style,
    }

    if (hasImage) {
      content = <img src={src} alt={alt} {...imgBaseProps} />
    } else if (icon) {
      content = <div className={classesMap?.icon}>{icon}</div>
    } else if (children || getAttribute('alt', imgBaseProps, alt)) {
      content = children || getCaptionFromAlt({ alt, imgProps: imgBaseProps })
    } else {
      content = <div className={classesMap?.fallback}>{fallback}</div>
    }

    return (
      <div ref={ref} {...rest} style={inlineStyle} className={avatarClassName}>
        {content}
      </div>
    )
  }
)
