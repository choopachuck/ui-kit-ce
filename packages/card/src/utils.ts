import { ButtonAriaActionEventHandler } from '@v-uik/hooks'
import { isSelectable, isClickable, CardProps, KindProps } from './types'

export const getInputProps = (
  props: KindProps
): React.HTMLProps<HTMLInputElement> | undefined => {
  if (!isSelectable(props)) {
    return undefined
  }

  return {
    checked: props.checked,
    onChange: props?.onChange,
    name: props.name,
    ...(props?.inputProps ?? {}),
  }
}

export const getClickableProps = (
  props: KindProps
):
  | (Omit<React.HTMLProps<HTMLDivElement>, 'onClick'> & {
      onClick?: ButtonAriaActionEventHandler<HTMLDivElement>
    })
  | undefined => {
  if (!isClickable(props)) {
    return undefined
  }

  return {
    ...props.buttonProps,
    onClick: props?.onClick,
  }
}

export const getOmittedParams = (props: CardProps) => {
  if (isSelectable(props)) {
    return ['checked', 'onChange', 'inputProps', 'kind', 'disabled']
  } else if (isClickable(props)) {
    return ['kind', 'buttonProps', 'onClick', 'disabled']
  }

  return ['kind']
}
