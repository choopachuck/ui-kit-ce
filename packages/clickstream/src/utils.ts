import {
  ClickStreamElementType,
  BaseElementType,
  InputElementType,
  ClickStreamElementProperty,
} from './types'
import {
  CONNECTION_TYPES,
  CLICK_STREAM_GENERIC_ATTRIBUTES,
  ALIGNMENT,
  MINUTES_IN_HOUR,
  MS_IN_MINUTE,
} from './constants'

declare global {
  interface Navigator {
    getBattery: <T>() => Promise<Record<string, T>>
    browserLanguage: string
    userLanguage: string
    systemLanguage: string
    connection?: {
      type?: keyof typeof CONNECTION_TYPES
    }
  }
}

export const isForbiddenForElement = (
  element: HTMLElement,
  eventType: string
): boolean => {
  if (
    element.hasAttribute(CLICK_STREAM_GENERIC_ATTRIBUTES.OFF) ||
    getTagName(element) === 'body'
  ) {
    return true
  }

  if (eventType === 'click') {
    if (
      element.hasAttribute('aria-disabled') &&
      element.getAttribute('aria-disabled') === 'true'
    ) {
      return true
    }

    if (
      isInput(element) ||
      isImg(element) ||
      element.hasAttribute('role') ||
      isSvg(element) ||
      isButton(element)
    ) {
      return false
    }

    return !element.innerText
  }

  return false
}

export const getClickEventElement = (
  event: Event
): [ClickStreamElementType, HTMLElement] => {
  const eventsPath = event.composedPath()
  const fallbackElements: Partial<
    Record<BaseElementType, [ClickStreamElementType, HTMLElement]>
  > = {}
  const defaultElement = [
    getTagName(eventsPath[0] as HTMLElement),
    eventsPath[0],
  ] as [ClickStreamElementType, HTMLElement]

  for (let i = 0; i <= eventsPath.length; i++) {
    const element = eventsPath[i] as HTMLElement
    if (isLabel(element)) {
      return [getTagName(element), element]
    }
    if (isInput(element)) {
      const input = element as HTMLInputElement

      return [input.type as ClickStreamElementType, input]
    }
    if (isButton(element)) {
      return [
        (element as HTMLButtonElement).type as ClickStreamElementType,
        element,
      ]
    }
    if (element instanceof HTMLLinkElement) {
      return ['link', element]
    }
    if (element instanceof HTMLAnchorElement) {
      return ['anchor', element]
    }
    if (element instanceof HTMLElement && element.hasAttribute('role')) {
      return [getTagName(element), element]
    }

    if (isSvg(element)) {
      fallbackElements.svg = [getTagName(element), element]
    }
  }

  return fallbackElements.svg || defaultElement
}

export const getChangeEventElement = (
  event: Event
): [ClickStreamElementType, HTMLInputElement | HTMLTextAreaElement] => {
  const element = event.target as HTMLInputElement | HTMLTextAreaElement
  const tagName = getTagName(element)

  if (tagName === 'select') {
    return [tagName as ClickStreamElementType, element]
  }

  if (element instanceof HTMLTextAreaElement) {
    return ['textarea', element]
  }

  return [element.type as InputElementType, element]
}

export const getElementAttributes = (
  element: HTMLElement
): ClickStreamElementProperty[] => {
  const attributes: ClickStreamElementProperty[] = []
  for (let i = 0, atts = element.attributes, n = atts.length; i < n; i++) {
    attributes.push({ key: atts[i].nodeName, value: String(atts[i].nodeValue) })
  }

  return attributes
}

export const isInput = (element: HTMLElement): boolean =>
  element instanceof HTMLInputElement || element instanceof HTMLTextAreaElement

export const isImg = (element: HTMLElement): boolean =>
  element instanceof HTMLImageElement

export const isSvg = (element: HTMLElement): boolean =>
  element instanceof SVGElement

export const isButton = (element: HTMLElement): boolean =>
  element instanceof HTMLButtonElement

export const isCheckboxOrRadio = (element: HTMLInputElement): boolean =>
  ['checkbox', 'radio'].includes(element.type)

export const isLabel = (element: HTMLElement): boolean =>
  element instanceof HTMLLabelElement

export const getTagName = (element: HTMLElement): BaseElementType =>
  element?.tagName?.toLowerCase() as BaseElementType

export const getElementValue = (
  element: HTMLElement
): string | boolean | null => {
  if (isInput(element)) {
    const input = element as HTMLInputElement
    const value = isCheckboxOrRadio(input) ? input.checked : input.value

    return value
  }

  if (isImg(element)) {
    return element.getAttribute('src')
  }

  return (
    element.getAttribute('aria-label') ||
    element.innerText ||
    element.textContent ||
    ''
  )
}

export const getApplicationLanguage = (): string =>
  globalThis.document?.documentElement.lang.substring(0, 16) || 'ru'

export const getBatteryLevel = async (): Promise<string> => {
  try {
    // Поддерживается не всеми браузерами
    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition, @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-assignment
    const battery = (await globalThis.navigator?.getBattery<number>()) || {}

    const { level } = battery

    if (typeof level === 'number') {
      return String(100 * level)
    }

    return '0'
  } catch (error) {
    return '0'
  }
}

export const getBrowser = (): string => globalThis.navigator?.userAgent

export const getConnectionType = (): string =>
  CONNECTION_TYPES[
    (globalThis.navigator?.connection?.type as keyof typeof CONNECTION_TYPES) ??
      'unknown'
  ]

export const getPageName = (): string => {
  if (!globalThis.document) {
    return ''
  }

  return globalThis.document.title
}

export const getUrl = (): string => {
  if (!window?.location?.href) {
    return ''
  }

  return window.location.href
}

export const getScreenSize = (): string =>
  globalThis.screen
    ? `${globalThis.screen.width}x${globalThis.screen.height}`
    : '0x0'

export const getSystemLanguage = (): string =>
  globalThis.navigator?.language ||
  globalThis.navigator?.userLanguage ||
  globalThis.navigator?.browserLanguage ||
  globalThis.navigator?.systemLanguage ||
  'ru'

const padStart = (value: number): number | string => {
  return value < ALIGNMENT ? `0${value}` : value
}

const formatOffset = (offset: number): string => {
  return `${offset < 0 ? '+' : '-'}${padStart(
    Math.abs(Math.floor(offset / MINUTES_IN_HOUR))
  )}:${padStart(Math.abs(offset % MINUTES_IN_HOUR))}`
}

export const getISOTime = (): string => {
  const date = new Date()
  const correctedDate = new Date(
    date.valueOf() - date.getTimezoneOffset() * MS_IN_MINUTE
  )
  const [correctedShortDate] = correctedDate.toISOString().split('Z')

  return `${correctedShortDate}${formatOffset(
    correctedDate.getTimezoneOffset()
  )}`
}
