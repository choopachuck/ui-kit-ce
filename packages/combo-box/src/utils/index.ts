export * from './utils'
export * from './createFilter'
