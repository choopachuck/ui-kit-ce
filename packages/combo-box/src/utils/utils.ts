import { Accessors, MultiValue, Options, SingleValue } from '../interfaces'

export const valueTernary = <Option>(
  isMulti: boolean,
  multiValue: MultiValue<Option>,
  singleValue: SingleValue<Option>
): SingleValue<Option> | MultiValue<Option> => {
  return isMulti ? multiValue : singleValue
}

export const isArrayOfStrings = (value: unknown): value is string[] => {
  return (
    Array.isArray(value) &&
    !!value.length &&
    value.every((item) => typeof item === 'string')
  )
}

export const formatValueToOption = <Option>(
  valueProp: string | string[] | Option | Options<Option> | undefined,
  options: Options<Option>,
  getOptionValue: (value: Option) => string
): Options<Option> => {
  if (valueProp === undefined || valueProp === null) {
    return []
  }

  if (isArrayOfStrings(valueProp)) {
    return valueProp
      .map((item) => {
        return options.find((option) => getOptionValue(option) === item) ?? null
      })
      .filter((option) => !!option) as Options<Option>
  }

  if (typeof valueProp === 'string' && valueProp.length > 0) {
    const findOption = options.find(
      (option) => getOptionValue(option) === valueProp
    )

    return findOption ? [findOption] : []
  }

  if (Array.isArray(valueProp)) {
    return valueProp
  }

  if (!Array.isArray(valueProp) && typeof valueProp !== 'string') {
    return [valueProp]
  }

  return []
}

export const cleanFullValue = <Option>(
  fullValue: MultiValue<Option> | SingleValue<Option> | undefined,
  newOption: Option,
  { getOptionValue }: Pick<Accessors<Option>, 'getOptionValue'>
): Options<Option> => {
  if (Array.isArray(fullValue)) {
    return fullValue.filter(
      (x) => getOptionValue(x) !== getOptionValue(newOption)
    )
  }

  return []
}

export const cleanValue = <Option>(
  value: SingleValue<Option> | MultiValue<Option> | undefined
): Options<Option> => {
  if (Array.isArray(value)) {
    return value.filter(Boolean) as Options<Option>
  }
  if (typeof value === 'object' && value !== null) {
    return [value] as Options<Option>
  }

  return []
}

export const isOption = <Option>(
  option: SingleValue<Option> | SingleValue<string>
): option is Option => {
  return typeof option !== 'string'
}

export const toOption = <Option>(
  option: SingleValue<Option> | SingleValue<string>
): Option => {
  if (isOption(option)) {
    return option
  } else {
    return {
      label: option,
      value: option,
    } as unknown as Option
  }
}

export const toString = <Option>(
  option: Option | string,
  getOptionLabel: (option: Option) => string
): string => {
  if (isOption(option)) {
    return getOptionLabel(option)
  } else {
    return option
  }
}

export function isCounter(limitTags?: number | null): boolean {
  return limitTags === 0 || limitTags === null
}
