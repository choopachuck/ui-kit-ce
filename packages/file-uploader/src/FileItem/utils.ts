import { ElementSizeType } from '@v-uik/common'

export const getIconSize = (size: ElementSizeType) => ({
  width: size === 'sm' ? 16 : 24,
  height: size === 'sm' ? 16 : 24,
})
