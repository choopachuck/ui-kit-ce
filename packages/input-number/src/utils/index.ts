export { createList } from './createList'
export type { Item } from './createList'
export * from './formatPastedValue'
export * from './filterPredicate'
