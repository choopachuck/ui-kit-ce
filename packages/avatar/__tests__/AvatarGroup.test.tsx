import * as React from 'react'
import { render } from '@testing-library/react'
import { light } from '@v-uik/theme'
import { shuffle } from '@v-uik/utils'
import {
  AvatarGroup,
  AvatarGroupProps,
  AvatarProps,
  AvatarSize,
  AvatarKind,
  AvatarGroupSize,
} from '../src'
import {
  avatarTestData,
  avatarGroupSizesAsArray,
  avatarKindsAsArray,
} from './testData'
import Extra from './testComponents/Extra'
import Avatar from './testComponents/Avatar'

jest.mock('./testComponents/Extra', () => {
  return jest.fn(() => null)
})

jest.mock('./testComponents/Avatar', () => {
  return jest.fn(() => null)
})

it.each(avatarGroupSizesAsArray)(
  'avatar group pass size property correctly (%p)',
  (size) => {
    const items = [0, 1, 2]
    const { getByTitle } = render(
      <AvatarGroup
        size={size}
        items={items.map((id) => ({ title: `${size}${id}` }))}
      />
    )

    items.forEach((id) => {
      expect(getByTitle(`${size}${id}`).getAttribute('class')).toContain(
        avatarTestData.size[size].className
      )
    })
  }
)

it.each(avatarKindsAsArray)(
  'avatar group pass kind property correctly (%p)',
  (kind) => {
    const items = [0, 1, 2]
    const { getByTitle } = render(
      <AvatarGroup
        kind={kind}
        items={items.map((id) => ({ title: `${kind}${id}` }))}
      />
    )

    items.forEach((id) => {
      expect(getByTitle(`${kind}${id}`).getAttribute('class')).toContain(
        avatarTestData.kind[kind].className
      )
    })
  }
)

it('avatar group pass withBorder property correctly', () => {
  const items = [0, 1, 2]
  const { getByTitle } = render(
    <AvatarGroup withBorder items={items.map((id) => ({ title: `${id}` }))} />
  )

  items.forEach((id) => {
    expect(getByTitle(`${id}`).getAttribute('class')).toContain(
      avatarTestData.withBorder.className
    )
  })
})

it('avatar group pass withShadow property correctly', () => {
  const items = [0, 1, 2]
  const { getByTitle } = render(
    <AvatarGroup withShadow items={items.map((id) => ({ title: `${id}` }))} />
  )

  items.forEach((id) => {
    expect(getByTitle(`${id}`).getAttribute('class')).toContain(
      avatarTestData.withShadow.className
    )
  })
})

it.each(avatarGroupSizesAsArray)(
  'avatar group gap changes via size property (%p)',
  (size) => {
    const items = [0, 1, 2]
    const { getByTitle } = render(
      <AvatarGroup
        size={size}
        items={items.map((id) => ({ title: `${size}${id}` }))}
      />
    )

    items.forEach((id) => {
      expect(getByTitle(`${size}${id}`)).toHaveStyle({
        ...avatarTestData.avatarGroup.size[size].style,
      })
    })
  }
)

it('avatar group gap changes via gap property', () => {
  const items = [0, 1, 2]
  const gap = 15
  const { getByTitle } = render(
    <AvatarGroup gap={gap} items={items.map((id) => ({ title: `${id}` }))} />
  )

  items.forEach((id) => {
    expect(getByTitle(`${id}`)).toHaveStyle({ marginLeft: gap * -1 })
  })
})

it('avatar group hides overflow avatars and shows count of hidden avatars', () => {
  const max = 5
  const total = 10
  const { container } = render(
    <AvatarGroup
      max={max}
      items={Array.from({ length: total }).map((_, index) => ({
        title: `${index}`,
      }))}
    />
  )

  const avatars = container.querySelectorAll(
    `[class^="${avatarTestData.avatarClassName}"]`
  )

  expect(avatars.length).toEqual(max + 1)
  expect(avatars[max].innerHTML).toContain(`+${total - max}`)
})

it('avatar group not shows count of hidden avatars if its not overflow', () => {
  const max = 11
  const total = 10
  const { container } = render(
    <AvatarGroup
      max={max}
      items={Array.from({ length: total }).map((_, index) => ({
        title: `${index}`,
      }))}
    />
  )

  const avatars = container.querySelectorAll(
    `[class^="${avatarTestData.avatarClassName}"]`
  )

  expect(avatars.length).toEqual(total)
  expect(avatars[avatars.length - 1].innerHTML).not.toContain(`+${total - max}`)
})

it('avatar group shows extra with showExtra=true', () => {
  const total = 5

  const { container } = render(
    <AvatarGroup
      showExtra
      items={Array.from({ length: total }).map((_, index) => ({
        title: `${index}`,
      }))}
    />
  )

  const avatars = container.querySelectorAll(
    `[class^="${avatarTestData.avatarClassName}"]`
  )

  expect(avatars.length).toEqual(total + 1)
})

it('avatar group renders custom extra correctly', () => {
  const max = 5
  const total = 6

  const avatarProps: AvatarProps = {
    kind: AvatarKind.rounded,
    size: AvatarSize.sm,
    color: 'red',
    children: 'UN',
    alt: 'User Name',
    src: './broken-image',
  }

  const avatarGroupProps: AvatarGroupProps = {
    kind: AvatarKind.circle,
    size: AvatarGroupSize.md,
    max,
    gap: 6,
    showExtra: false,
    items: Array.from({ length: total }).map(() => ({ ...avatarProps })),
  }

  render(<AvatarGroup {...avatarGroupProps} components={{ Extra }} />)

  expect(Extra).toBeCalledTimes(1)
  expect(Extra).toMatchSnapshot()
})

it('avatar group renders custom avatar correctly', () => {
  render(
    <AvatarGroup
      items={[{ src: './broken_image.png', alt: 'UN' }]}
      components={{ Avatar }}
    />
  )

  expect(Avatar).toBeCalledTimes(1)
  expect(Avatar).toMatchSnapshot()
})

it('avatar group changes background color of avatars via property coloring="static" correclty', () => {
  const backgroundColors = [
    light.ref.palette.amber50,
    light.ref.palette.arctic50,
    light.ref.palette.blue50,
  ]
  const items = [0, 1, 2]
  const { getByTitle } = render(
    <AvatarGroup
      coloring="static"
      items={items.map((id) => ({ title: `${id}` }))}
    />
  )

  items.forEach((id) => {
    expect(getByTitle(`${id}`)).toHaveStyle({
      backgroundColor: backgroundColors[id],
    })
  })
})

it('avatar group changes background color of avatars via property coloring="static" correclty with custom palette', () => {
  const backgroundColors = ['red', 'green', 'blue']
  const textColorsByBackgroundColors = {
    [backgroundColors[0]]: 'rgb(255, 255, 255)',
    [backgroundColors[1]]: 'rgb(255, 255, 255)',
    [backgroundColors[2]]: 'rgb(0, 0, 0)',
  }
  const palette = [
    {
      textColor: textColorsByBackgroundColors[backgroundColors[0]],
      backgroundColor: backgroundColors[0],
    },
    {
      textColor: textColorsByBackgroundColors[backgroundColors[1]],
      backgroundColor: backgroundColors[1],
    },
    {
      textColor: textColorsByBackgroundColors[backgroundColors[2]],
      backgroundColor: backgroundColors[2],
    },
  ]
  const items = [0, 1, 2]
  const { getByTitle } = render(
    <AvatarGroup
      coloring="static"
      getPalette={() => palette}
      items={items.map((id) => ({ title: `${id}` }))}
    />
  )

  items.forEach((id) => {
    expect(getByTitle(`${id}`)).toHaveStyle({
      backgroundColor: backgroundColors[id],
      color: textColorsByBackgroundColors[backgroundColors[id]],
    })
  })
})

it('avatar group changes background color of avatars via property coloring="random" correctly', () => {
  const backgroundColors = ['red', 'green', 'blue']
  const textColorsByBackgroundColors = {
    [backgroundColors[0]]: 'rgb(255, 255, 255)',
    [backgroundColors[1]]: 'rgb(255, 255, 255)',
    [backgroundColors[2]]: 'rgb(0, 0, 0)',
  }
  const palette = [
    {
      textColor: textColorsByBackgroundColors[backgroundColors[0]],
      backgroundColor: backgroundColors[0],
    },
    {
      textColor: textColorsByBackgroundColors[backgroundColors[1]],
      backgroundColor: backgroundColors[1],
    },
    {
      textColor: textColorsByBackgroundColors[backgroundColors[2]],
      backgroundColor: backgroundColors[2],
    },
  ]
  const items = [0, 1, 2]
  const { getByTitle } = render(
    <AvatarGroup
      coloring="random"
      getPalette={() => shuffle(palette)}
      items={items.map((id) => ({ title: `${id}` }))}
    />
  )

  items.forEach((id) => {
    expect(backgroundColors).toContain(
      getByTitle(`${id}`).style.backgroundColor
    )
    expect(getByTitle(`${id}`).style.color).toEqual(
      textColorsByBackgroundColors[getByTitle(`${id}`).style.backgroundColor]
    )
  })
})

it('avatar group with coloring and small palette changes background color of avatars correctly', () => {
  const backgroundColors = ['red', 'green', 'blue']
  const textColorsByBackgroundColors = {
    [backgroundColors[0]]: 'rgb(255, 255, 255)',
    [backgroundColors[1]]: 'rgb(255, 255, 255)',
    [backgroundColors[2]]: 'rgb(0, 0, 0)',
  }
  const palette = [
    {
      textColor: textColorsByBackgroundColors[backgroundColors[0]],
      backgroundColor: backgroundColors[0],
    },
    {
      textColor: textColorsByBackgroundColors[backgroundColors[1]],
      backgroundColor: backgroundColors[1],
    },
    {
      textColor: textColorsByBackgroundColors[backgroundColors[2]],
      backgroundColor: backgroundColors[2],
    },
  ]
  const items = [0, 1, 2, 3, 4, 5]
  const { getByTitle } = render(
    <AvatarGroup
      coloring="static"
      getPalette={() => palette}
      items={items.map((id) => ({ title: `${id}` }))}
    />
  )

  let paletteId = 0
  items.forEach((id) => {
    if (paletteId === backgroundColors.length) {
      paletteId = 0
    }
    expect(getByTitle(`${id}`)).toHaveStyle({
      backgroundColor: backgroundColors[paletteId],
      color: textColorsByBackgroundColors[backgroundColors[paletteId]],
    })
    paletteId++
  })
})
