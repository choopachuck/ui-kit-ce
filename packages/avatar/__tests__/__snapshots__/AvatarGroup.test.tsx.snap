// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`avatar group renders custom avatar correctly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "alt": "UN",
        "className": "avatarGroup-item-0-2-1",
        "src": "./broken_image.png",
      },
      Object {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": null,
    },
  ],
}
`;

exports[`avatar group renders custom extra correctly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "avatarProps": Object {
          "classes": Object {
            "root": "avatarGroup-item-0-2-1",
          },
          "kind": "circle",
          "size": "md",
          "style": Object {
            "marginLeft": -6,
          },
        },
        "children": "+1",
        "classes": Object {
          "extraSmall": "avatarGroup-extraSmall-0-2-3",
          "item": "avatarGroup-item-0-2-1",
          "medium": "avatarGroup-medium-0-2-5",
          "root": "avatarGroup-root-0-2-2",
          "small": "avatarGroup-small-0-2-4",
        },
        "gap": 6,
        "hiddenItems": Array [
          Object {
            "alt": "User Name",
            "children": "UN",
            "className": "",
            "color": "red",
            "kind": "rounded",
            "size": "sm",
            "src": "./broken-image",
          },
        ],
        "kind": "circle",
        "max": 5,
        "showExtra": false,
        "size": "md",
        "visibleItems": Array [
          Object {
            "alt": "User Name",
            "children": "UN",
            "className": "avatarGroup-item-0-2-1",
            "color": "red",
            "kind": "rounded",
            "size": "sm",
            "src": "./broken-image",
          },
          Object {
            "alt": "User Name",
            "children": "UN",
            "className": "avatarGroup-item-0-2-1",
            "color": "red",
            "kind": "rounded",
            "size": "sm",
            "src": "./broken-image",
            "style": Object {
              "marginLeft": -6,
            },
          },
          Object {
            "alt": "User Name",
            "children": "UN",
            "className": "avatarGroup-item-0-2-1",
            "color": "red",
            "kind": "rounded",
            "size": "sm",
            "src": "./broken-image",
            "style": Object {
              "marginLeft": -6,
            },
          },
          Object {
            "alt": "User Name",
            "children": "UN",
            "className": "avatarGroup-item-0-2-1",
            "color": "red",
            "kind": "rounded",
            "size": "sm",
            "src": "./broken-image",
            "style": Object {
              "marginLeft": -6,
            },
          },
          Object {
            "alt": "User Name",
            "children": "UN",
            "className": "avatarGroup-item-0-2-1",
            "color": "red",
            "kind": "rounded",
            "size": "sm",
            "src": "./broken-image",
            "style": Object {
              "marginLeft": -6,
            },
          },
        ],
      },
      Object {},
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": null,
    },
  ],
}
`;
